sessions/04-using-delay-distributions-to-model-the-data-generating-process-of-an-epidemic/index.qmd

```diff
@@ -2,3 +2,291 @@
+## delay distribution
+## example: convolve_with_delay(c(10, 14, 10, 10), c(0.1, 0.6, 0.3))
+convolve_with_delay <- function(ts, delay) {
+  max_delay <- length(delay) - 1 ## subtract one because zero-indexed
+  convolved <- vapply(seq_along(ts), \(i) {
+    ## get vector of infections (14-day window)
+    first_index <- max(1, i - max_delay)
+    ts_segment <- ts[seq(first_index, i)]
+    ## take reverse of pmf and reverse if needed
+    pmf <- rev(delay)[seq_len(i - first_index + 1)]
+    ## convolve with delay distribution
+    ret <- sum(ts_segment * pmf)
+    return(round(ret))
+  }, numeric(1))
+  return(convolved)
+}
+```
+
+::: {.callout-tip}
+## Take 5 minutes
+Try to understand the `convolve_with_delay()` function above.
+Try it with a few different time series and delay distributions.
+How would you create the time series of infections using the discretised gamma distribution created above and saved in `gamma_pmf`?
+:::
+
+::: {.callout-note collapse="true"]}
+```{r applied_convolution}
+onsets <- convolve_with_delay(inf_time_series$n, gamma_pmf)
+```
+:::
+
+We can plot these symptom onsets
+```{r convolution_plot}
+combined <- inf_time_series |>
+  rename(date = infection_date, infections = n) |>
+  mutate(onsets = onsets)
+ggplot(combined, aes(x = date, y = onsets)) +
+  geom_bar(stat = "identity")
+```
+
+Do they look similar to the plot of symptom onsets in the previous session?
+
+## Observation uncertainty
+
+Usually not all data are perfectly observed.
+The counting process itself generally introduces a natural form of error that we can, once again, model with a probability distribution.
+For modelling counts, a common choice is the Poisson distribution.
+We can use this to generate uncertainty around our convolved data.
+
+```{r uncertain}
+onsets <- onsets |>
+  mutate(observed = rpois(n(), onsets))
+```
+
+Does a plot of these observations look more like the data from the previous session than the convolution plotted above?
+
+# Estimating a time series of infection
+
+We now invert the procedure of above, i.e. instead of convolving infects and adding uncertainty to simulate observed symptom onsets, we start with observed simulated onsets to estimate the number of infections, given an incubation period.
+Mathematically this amounts to a _deconvolution_ of the time series, i.e. the reverse process to convolution.
+
+```{r stan_deconvolve}
+mod <- cmdstan_model(
+  here(
+    "sessions",
+    paste0(
+      "04-using-delay-distributions-to-model-the-data-generating-process-of-",
+      "an-epidemic"
+    ),
+    "estimate_infections.stan"
+  )
+)
+mod$print(line_numbers = TRUE)
+```
+
+::: {.callout-tip}
+## Take 10 minutes
+Familarise yourself with the model above.
+Unlike before there is now a `functions` block at the beginning of the model (lines 1-33), where we define two functions, `discretise_gamma()` (line 2) and `convolve_with_delay()` (line 18).
```

Comment:
For simplicity here I thought we would use the `pmf` calculated above? I don't mind hugely either way but this approach does add some model complexity for people to read through and maybe distracts from the core point? Or do you think teaching people about uncertainty propagation is worth it?

Reply:
Good idea.